File: src/main/java/com/baulsupp/oksocial/services/fitbit/FitbitAuthInterceptor.java
package com.baulsupp.oksocial.services.fitbit;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.authenticator.AuthUtil;
import com.baulsupp.oksocial.authenticator.JsonCredentialsValidator;
import com.baulsupp.oksocial.authenticator.ValidatedCredentials;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition;
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token;
import com.baulsupp.oksocial.output.OutputHandler;
import com.baulsupp.oksocial.secrets.Secrets;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.Future;
import okhttp3.Credentials;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import static com.baulsupp.oksocial.authenticator.JsonCredentialsValidator.fieldExtractor;

public class FitbitAuthInterceptor implements AuthInterceptor<Oauth2Token> {
  @Override public Oauth2ServiceDefinition serviceDefinition() {
    return new Oauth2ServiceDefinition("api.fitbit.com", "Fitbit API", "fitbit",
        "https://dev.fitbit.com/docs/", "https://dev.fitbit.com/apps/");
  }

  @Override public Response intercept(Interceptor.Chain chain, Oauth2Token credentials)
      throws IOException {
    Request request = chain.request();

    String token = credentials.accessToken;

    request =
        request.newBuilder().addHeader("Authorization", "Bearer " + token).build();

    return chain.proceed(request);
  }

  @Override public Oauth2Token authorize(OkHttpClient client, OutputHandler outputHandler,
      List<String> authArguments) throws IOException {
    System.err.println("Authorising Fitbit API");

    String clientId =
        Secrets.prompt("Fitbit Client Id", "fitbit.clientId", "", false);
    String clientSecret =
        Secrets.prompt("Fitbit Client Secret", "fitbit.clientSecret", "", true);
    Set<String> scopes = Secrets.promptArray("Scopes", "fitbit.scopes", FitbitUtil.SCOPES);

    return FitbitAuthCodeFlow.login(client, outputHandler, clientId, clientSecret, scopes);
  }

  @Override public Future<Optional<ValidatedCredentials>> validate(OkHttpClient client,
      Request.Builder requestBuilder, Oauth2Token credentials) throws IOException {
    return new JsonCredentialsValidator(
        FitbitUtil.apiRequest("/1/user/-/profile.json", requestBuilder), this::getName).validate(
        client);
  }

  @Override public boolean canRenew(Oauth2Token credentials) {
    return credentials.refreshToken.isPresent()
        && credentials.clientId.isPresent()
        && credentials.clientSecret.isPresent();
  }

  @Override
  public Optional<Oauth2Token> renew(OkHttpClient client, Oauth2Token credentials)
      throws IOException {

    RequestBody body = RequestBody.create(MediaType.parse("application/json"),
        "{\"grant_type\": \"refresh_token\", \"refresh_token\": \""
            + credentials.refreshToken.get() + "\"}");
    String basic = Credentials.basic(credentials.clientId.get(), credentials.clientSecret.get());
    Request request =
        new Request.Builder().url("https://api.fitbit.com/oauth2/token")
            .post(body)
            .header("Authorization", basic)
            .build();

    Map<String, Object> responseMap = AuthUtil.makeJsonMapRequest(client, request);

    return Optional.of(new Oauth2Token((String) responseMap.get("access_token"),
        credentials.refreshToken.get(), credentials.clientId.get(),
        credentials.clientSecret.get()));
  }

  private String getName(Map<String, Object> map) {
    Map<String, Object> user = (Map<String, Object>) map.get("user");

    return (String) user.get("fullName");
  }

  @Override public Set<String> hosts() {
    return FitbitUtil.API_HOSTS;
  }
}

File: src/main/java/com/baulsupp/oksocial/services/fitbit/FitbitAuthInterceptor.kt
package com.baulsupp.oksocial.services.fitbit

import com.baulsupp.oksocial.authenticator.AuthInterceptor
import com.baulsupp.oksocial.authenticator.AuthUtil
import com.baulsupp.oksocial.authenticator.JsonCredentialsValidator
import com.baulsupp.oksocial.authenticator.ValidatedCredentials
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2ServiceDefinition
import com.baulsupp.oksocial.authenticator.oauth2.Oauth2Token
import com.baulsupp.oksocial.secrets.Secrets
import com.baulsupp.oksocial.output.OutputHandler
import java.io.IOException
import java.util.Optional
import java.util.concurrent.Future
import okhttp3.Credentials
import okhttp3.FormBody
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response

class FitbitAuthInterceptor : AuthInterceptor<Oauth2Token> {
    override fun serviceDefinition(): Oauth2ServiceDefinition {
        return Oauth2ServiceDefinition("api.fitbit.com", "Fitbit API", "fitbit",
                "https://dev.fitbit.com/docs/", "https://dev.fitbit.com/apps/")
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain, credentials: Oauth2Token): Response {
        var request = chain.request()

        val token = credentials.accessToken

        request = request.newBuilder().addHeader("Authorization", "Bearer " + token).build()

        return chain.proceed(request)
    }

    @Throws(IOException::class)
    fun authorize(client: OkHttpClient, outputHandler: OutputHandler<*>,
                  authArguments: List<String>): Oauth2Token {
        System.err.println("Authorising Fitbit API")

        val clientId = Secrets.prompt("Fitbit Client Id", "fitbit.clientId", "", false)
        val clientSecret = Secrets.prompt("Fitbit Client Secret", "fitbit.clientSecret", "", true)
        val scopes = Secrets.promptArray("Scopes", "fitbit.scopes", FitbitUtil.SCOPES)

        return FitbitAuthCodeFlow.login(client, outputHandler, clientId, clientSecret, scopes)
    }

    @Throws(IOException::class)
    override fun validate(client: OkHttpClient,
                          requestBuilder: Request.Builder, credentials: Oauth2Token): Future<Optional<ValidatedCredentials>> {
        return JsonCredentialsValidator(
                FitbitUtil.apiRequest("/1/user/-/profile.json", requestBuilder), Function<Map<String, Any>, String> { this.getName(it) }).validate(
                client)
    }

    override fun canRenew(credentials: Oauth2Token): Boolean {
        return credentials.refreshToken.isPresent
                && credentials.clientId.isPresent
                && credentials.clientSecret.isPresent
    }

    @Throws(IOException::class)
    override fun renew(client: OkHttpClient, credentials: Oauth2Token): Optional<Oauth2Token> {
        val body = FormBody.Builder().add("grant_type", "refresh_token")
                .add("refresh_token", credentials.refreshToken.get())
                .build()
        val basic = Credentials.basic(credentials.clientId.get(), credentials.clientSecret.get())
        val request = Request.Builder().url("https://api.fitbit.com/oauth2/token")
                .post(body)
                .header("Authorization", basic)
                .build()

        val responseMap = AuthUtil.makeJsonMapRequest(client, request)

        return Optional.of(Oauth2Token(responseMap["access_token"] as String,
                credentials.refreshToken.get(), credentials.clientId.get(),
                credentials.clientSecret.get()))
    }

    private fun getName(map: Map<String, Any>): String {
        val user = map["user"] as Map<String, Any>

        return user["fullName"] as String
    }

    override fun hosts(): Set<String> {
        return FitbitUtil.API_HOSTS
    }
}
